src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,555 @@
+
+        $pad = clone $this;
+        $pad->string = $padStr;
+
+        return $this->pad($length, $pad, STR_PAD_RIGHT);
+    }
+
+    /**
+     * @return static
+     */
+    abstract public function prepend(string $prefix): self;
+
+    /**
+     * @return static
+     */
+    public function repeat(int $multiplier): self
+    {
+        $str = clone $this;
+        $str->string = str_repeat($str->string, $multiplier < 0 ? 0 : $multiplier);
+
+        return $str;
+    }
+
+    /**
+     * @return static
+     */
+    abstract public function replace(string $from, string $to): self;
+
+    /**
+     * @param string|callable $to
+     *
+     * @return static
+     */
+    public function replaceMatches(string $from, $to): self
+    {
+        if ('g' === $pattern[-1]) {
+            $pattern = substr($pattern, 0, -1);
+        }
+
+        if ($this->ignoreCase) {
+            $pattern .= 'i';
+        }
+
+        if (\is_array($to) || $to instanceof \Closure) {
+            $to = static function (array $m) use ($to): string {
+                $to = $to($m);
+
+                if ('' !== $to && (!\is_string($to) || !preg_match('//u', $to))) {
+                    throw new InvalidArgumentException('Replace callback must return a valid UTF-8 string.');
+                }
+
+                return $to;
+            };
+        } elseif ('' !== $to && !preg_match('//u', $to)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        } else {
+            $replace = 'preg_replace';
+        }
+
+        set_error_handler(static function ($t, $m) { throw new InvalidArgumentException($m); });
+
+        try {
+            if (null === $string = $replace($from.'u', $to, $this->string)) {
+                $lastError = preg_last_error();
+
+                foreach (get_defined_constants(true)['pcre'] as $k => $v) {
+                    if ($lastError === $v && '_ERROR' === substr($k, -6)) {
+                        throw new RuntimeException('Matching failed with '.$k.'.');
+                    }
+                }
+
+                throw new RuntimeException('Matching failed with unknown error code.');
+            }
+        } finally {
+            restore_error_handler();
+        }
+
+        $str = clone $this;
+        $str->string = $string;
+
+        return $str;
+    }
+
+    /**
+     * @return static[]
+     */
+    abstract public function split(string $delimiter, int $limit = null): array;
+
+    abstract public function startsWith(string $prefix): bool;
+
+    /**
+     * @return static
+     */
+    abstract public function substr(int $start = 0, int $length = null): self;
+
+    /**
+     * @return static
+     */
+    abstract public function substrReplace(string $replacement, int $start = 0, int $length = null): self;
+
+    /**
+     * @return static
+     */
+    abstract public function toFoldedCase(): self;
+
+    /**
+     * @return static
+     */
+    public function toLowerCase(): self
```

Comment:
```suggestion
    public function lowercase(): self
```

Reply:
toLowerCase is much more common in many libs and languages

Reply:
i think we should at least change this to `toLowercase()`, since [`lowercase` is consider 1 word](https://english.stackexchange.com/questions/59409/lowercase-lower-case-or-lower-case).

other language / standard library using `lowercase` instead of `lowerCase` or `lower_case` : 

- Rust : https://doc.rust-lang.org/std/string/struct.String.html#examples-91
- Hack : https://github.com/hhvm/hsl/blob/master/src/str/transform.php#L76-L87 

( same goes for `uppercase` )

Reply:
Funny you linked to languages that use snake case. Here are two others that use camel case (like we do):
Java, JavaScript: :)
https://docs.oracle.com/javase/7/docs/api/java/lang/String.html#toLowerCase(java.util.Locale)
https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/String/toLowerCase

Reply:
i actually linked language that consider "lowercase" to be 1 word, and use it as is, instead of "lower_case" :)

Reply:
@azjezz I looked for this and apparently, in English both "lower case" and "lowercase" are correct and commonly used (outside of programming context).

Reply:
Given that PHP method names are actually case insensitive, I suggest not spending hours on this then...